packages/shared/src/util/index.ts

```diff
@@ -19,16 +20,26 @@ export * from "./features";
 
 export function getAffectedEnvsForExperiment({
   experiment,
+  linkedFeatures,
 }: {
   experiment: ExperimentInterface | ExperimentInterfaceStringDates;
+  linkedFeatures?: LinkedFeatureInfo[];
 }): string[] {
   // Visual changesets are not environment-scoped, so it affects all of them
   if (experiment.hasVisualChangesets || experiment.hasURLRedirects)
     return ["__ALL__"];
 
-  // TODO: get actual environments for linked feature flags. We are being overly conservative here
-  if (experiment.linkedFeatures && experiment.linkedFeatures.length > 0) {
-    return ["__ALL__"];
+  if (linkedFeatures && linkedFeatures.length > 0) {
+    const envs: string[] = [];
+
+    linkedFeatures.forEach((linkedFeature) => {
+      for (const env in linkedFeature.environmentStates) {
+        if (linkedFeature.environmentStates[env] === "active") {
+          envs.push(env);
+        }
+      }
+    });
+    return envs;
   }
```

Comment:
Since the `linkedFeatures` argument is optional, should we have a fallback that returns `["__ALL__"]` if it's not passed in and `experiment.linkedFeatures` is non-empty?